clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/PubSubBasedAdminClientManager.java

```diff
@@ -148,68 +145,53 @@ public Future<Void> closeAdminClients() {
-                .recover(Future::failedFuture)
+                .onFailure(thr -> LOG.debug("admin client creation failed", thr))
                 .compose(client -> {
                     topicAdminClient = client;
                     return getOrCreateTopic(projectId, topic);
                 });
     }
 
     private Future<Void> getOrCreateTopic(final String projectId, final TopicName topic) {
-        if (topicExists(topic.toString())) {
-            return Future.succeededFuture();
-        }
-        final Topic createdTopic = topicAdminClient.createTopic(topic);
-        if (createdTopic == null) {
-            LOG.debug("Creating topic failed [topic: {}, projectId: {}]", topic.toString(), projectId);
-            return Future.failedFuture("Topic creation failed.");
-        }
-        topics.add(topic.toString());
-        return Future.succeededFuture();
-    }
-
-    private boolean topicExists(final String topicId) {
-        if (topics.contains(topicId)) {
-            return true;
-        }
-        final ProjectName projectName = ProjectName.of(projectId);
         try {
-            final ListTopicsPagedResponse pagedResponse = topicAdminClient.listTopics(projectName);
-            if (pagedResponse != null) {
-                for (Topic topic : pagedResponse.iterateAll()) {
-                    if (topic.getName().equals(topicId)) {
-                        topics.add(topicId);
-                        return true;
-                    }
-                }
+            final Topic createdTopic = topicAdminClient.createTopic(topic);
+            if (createdTopic == null) {
+                LOG.debug("Creating topic failed [topic: {}, projectId: {}]", topic, projectId);
+                return Future.failedFuture("Topic creation failed.");
             }
-            return false;
+            topics.add(topic.toString());
+            return Future.succeededFuture();
+        } catch (AlreadyExistsException ex) {
+            return Future.succeededFuture();
         } catch (ApiException e) {
-            LOG.error("Error listing topics on project {}", projectId, e);
-            return false;
+            LOG.debug("Error creating topic {} on project {}", topic, projectId, e);
+            return Future.failedFuture("Topic creation failed.");
         }
 
     }
 
     private Future<String> getOrCreateSubscription(final SubscriptionName subscription, final TopicName topic) {
+        if (subscriptions.contains(subscription.toString())) {
+            LOG.debug("Subscription {} already exists, continue", subscription);
+            return Future.succeededFuture(subscription.getSubscription());
+        }
         return subscriptionAdminClientCreator.get()
-                .onFailure(thr -> LOG.error("admin client creation failed", thr))
-                .recover(Future::failedFuture)
+                .onFailure(thr -> LOG.debug("admin client creation failed", thr))
```

Comment:
again, keeping a reference to the *createSubscriptionAdminClient()* in a field instead of invoking the method directly feels strange. Is there a particular reason for doing this?

Reply:
see above

Reply:
IMHO it does not make much sense to store the admin client in the *subscriptionAdminClient* field if the field is never being read ....

Reply:
good point, I changed the implementation to make use of the _subscriptionAdminClient_ in a meaningful way.